firebase-firestore/src/main/java/com/google/firebase/firestore/LoadBundleTask.java

```diff
@@ -15,183 +15,556 @@
+   * @see SuccessContinuation#then(ResultT)
+   */
   @NonNull
   @Override
   public <TContinuationResult> Task<TContinuationResult> onSuccessTask(
       @NonNull Executor executor,
       @NonNull
           SuccessContinuation<LoadBundleTaskProgress, TContinuationResult> successContinuation) {
-    return super.onSuccessTask(executor, successContinuation);
+    return delegate.onSuccessTask(executor, successContinuation);
+  }
+
+  /**
+   * Adds a listener that is called periodically while the LoadBundleTask executes.
+   *
+   * @return this Task
+   */
+  @NonNull
+  public LoadBundleTask addOnProgressListener(
+      @NonNull OnProgressListener<LoadBundleTaskProgress> listener) {
+    ManagedListener managedListener = new ManagedListener(/* executor= */ null, listener);
+    synchronized (lock) {
+      progressListenerQueue.add(managedListener);
+      managedListener.maybeRun(snapshot);
+    }
+    return this;
+  }
+
+  /**
+   * Adds a listener that is called periodically while the LoadBundleTask executes.
+   *
+   * @param executor the executor to use to call the listener
+   * @return this Task
+   */
+  @NonNull
+  public LoadBundleTask addOnProgressListener(
+      @NonNull Executor executor, @NonNull OnProgressListener<LoadBundleTaskProgress> listener) {
+    ManagedListener managedListener = new ManagedListener(executor, listener);
+    synchronized (lock) {
+      progressListenerQueue.add(managedListener);
+      managedListener.maybeRun(snapshot);
+    }
+    return this;
+  }
+
+  /**
+   * Adds a listener that is called periodically while the LoadBundleTask executes.
+   *
+   * @param activity When the supplied {@link Activity} stops, this listener will automatically be
+   *     removed.
+   * @return this Task
+   */
+  @NonNull
+  public LoadBundleTask addOnProgressListener(
+      @NonNull Activity activity, @NonNull OnProgressListener<LoadBundleTaskProgress> listener) {
+    ManagedListener managedListener = new ManagedListener(/* executor= */ null, listener);
+    synchronized (lock) {
+      progressListenerQueue.add(managedListener);
+      managedListener.maybeRun(snapshot);
+    }
+    ActivityLifecycleListener.getInstance()
+        .runOnActivityStopped(activity, listener, () -> removeOnProgressListener(listener));
+    return this;
+  }
+
+  /**
+   * Removes a listener.
+   *
+   * <p>To match the Android Task API and its usage across Firestore, this method is private.
+   */
+  private void removeOnProgressListener(
+      @NonNull OnProgressListener<LoadBundleTaskProgress> listener) {
+    synchronized (lock) {
+      progressListenerQueue.remove(new ManagedListener(/* executor= */ null, listener));
+    }
+  }
+
+  void setResult(@Nullable LoadBundleTaskProgress result) {
+    synchronized (lock) {
+      snapshot = result;
+      for (ManagedListener progressListener : progressListenerQueue) {
+        progressListener.maybeRun(result);
```

Comment:
I think this code will throw ConcurrentModificationException if the callback adds or removes a progress listener. Please add a test for this and verify that this scenario is handled correctly. This comment also applies to `setException()` and `updateProgress()`.